File: backend/F1Fantasy/Controllers/DriverController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.AspNetCore.RateLimiting;
using F1Fantasy.Models;
using F1Fantasy.Services;

namespace F1Fantasy.Controllers;

[Authorize]
[ApiController]
[Route("api/[controller]")]
[EnableRateLimiting("read")]
public class DriverController : ControllerBase
{
    private readonly DriverService _driverService;
    private readonly ILogger<DriverController> _logger;

    public DriverController(DriverService driverService, ILogger<DriverController> logger)
    {
        _driverService = driverService;
        _logger = logger;
    }

    [HttpGet]
    public async Task<ActionResult<IEnumerable<Driver>>> GetAllDrivers()
    {
        try
        {
            _logger.LogInformation("GET /api/driver - Fetching all drivers");
            var drivers = await _driverService.GetAllDriversAsync();
            _logger.LogInformation("Successfully retrieved {Count} total drivers", drivers.Count());
            return Ok(drivers);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error fetching all drivers");
            throw;
        }
    }

    [HttpGet("season/{season}")]
    public async Task<ActionResult<IEnumerable<Driver>>> GetDriversBySeason(string season)
    {
        try
        {
            _logger.LogInformation("GET /api/driver/season/{Season} - Fetching drivers for season", season);
            var drivers = await _driverService.GetDriversBySeasonAsync(season);
            _logger.LogInformation("Successfully retrieved {Count} drivers for season {Season}", drivers.Count(), season);
            return Ok(drivers);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error fetching drivers for season {Season}", season);
            throw;
        }
    }

    [HttpGet("{driverId}")]
    public async Task<ActionResult<Driver>> GetDriverById(string driverId)
    {
        try
        {
            _logger.LogInformation("GET /api/driver/{DriverId} - Fetching driver by ID", driverId);
            var driver = await _driverService.GetDriverByIdAsync(driverId);
            if (driver == null)
            {
                _logger.LogWarning("Driver not found: {DriverId}", driverId);
                return NotFound(new { message = $"Driver not found: {driverId}" });
            }
            _logger.LogInformation("Successfully retrieved driver {DriverId}", driverId);
            return Ok(driver);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error fetching driver {DriverId}", driverId);
            throw;
        }
    }

    [HttpGet("cached")]
    public async Task<ActionResult<IEnumerable<Driver>>> GetCachedDrivers()
    {
        try
        {
            _logger.LogInformation("GET /api/driver/cached - Fetching cached drivers");
            var drivers = await _driverService.GetCachedDriversAsync();
            
            if (!drivers.Any())
            {
                _logger.LogWarning("No cached drivers found. Database may be empty.");
                return Ok(new { 
                    message = "No cached drivers found. Try calling /api/driver/season/2024 first to populate the cache.",
                    drivers = drivers 
                });
            }
            
            _logger.LogInformation("Successfully retrieved {Count} cached drivers", drivers.Count());
            return Ok(drivers);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error fetching cached drivers");
            throw;
        }
    }
}

File: backendJava/f1fantasy/src/main/java/no/f1fantasy/controller/DriverController.java
package no.f1fantasy.controller;

import java.util.List;

import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import no.f1fantasy.entity.Driver;
import no.f1fantasy.service.DriverService;

@RestController
@RequestMapping("/api/driver")
@PreAuthorize("isAuthenticated()")
@SuppressWarnings("null")
public class DriverController {

    private final DriverService driverService;

    public DriverController(DriverService driverService) {
        this.driverService = driverService;
    }

    @GetMapping
    public ResponseEntity<?> getAllDrivers() {
        try {
            List<Driver> drivers = driverService.getAllDrivers();
            return ResponseEntity.ok(drivers);
        } catch (Exception ex) {
            return ResponseEntity.status(500).body(new ErrorResponse("Error fetching drivers"));
        }
    }

    @GetMapping("/season/{season}")
    public ResponseEntity<?> getDriversBySeason(@PathVariable String season) {
        try {
            List<Driver> drivers = driverService.getDriversBySeason(season);
            return ResponseEntity.ok(drivers);
        } catch (Exception ex) {
            return ResponseEntity.status(500).body(new ErrorResponse("Error fetching drivers for season: " + season));
        }
    }

    @GetMapping("/{driverId}")
    public ResponseEntity<?> getDriverById(@PathVariable String driverId) {
        try {
            java.util.Optional<Driver> driver = driverService.getDriverById(driverId);
            if (!driver.isPresent()) {
                return ResponseEntity.status(404).body(new ErrorResponse("Driver not found: " + driverId));
            }
            return ResponseEntity.ok(driver.get());
        } catch (Exception ex) {
            return ResponseEntity.status(500).body(new ErrorResponse("Error fetching driver: " + driverId));
        }
    }

    @GetMapping("/cached")
    public ResponseEntity<?> getCachedDrivers() {
        try {
            List<Driver> drivers = driverService.getAllDrivers();
            if (drivers.isEmpty()) {
                return ResponseEntity.ok(new CachedResponse("No cached drivers found", drivers));
            }
            return ResponseEntity.ok(drivers);
        } catch (Exception ex) {
            return ResponseEntity.status(500).body(new ErrorResponse("Error fetching cached drivers"));
        }
    }

    @SuppressWarnings ("unused")
    private static class ErrorResponse {
        public final String error;

        public ErrorResponse(String error) {
            this.error = error;
        }
    }

    @SuppressWarnings("unused")
    private static class CachedResponse {
        public final String message;
        public final List<?> drivers;

        public CachedResponse(String message, List<?> drivers) {
            this.message = message;
            this.drivers = drivers;
        }
    }
}
